File: session4/137_div2_educational/c.cpp
#include <bits/stdc++.h>
#define int long long
using namespace std;



int32_t main(){
    int test_num;cin>>test_num;
    for(int i=0;i<test_num;i++){
        int num;cin>>num;
        vector<pair<int,char>>arr(num);
        for(int j=0;j<num;j++){
            cin>>arr[j].second;
        }
        for(int j=0;j<num;j++){
            cin>>arr[j].first;
        }
        int result=0;
        for(int j=0;j<num-1;j++){
            if(arr[j].second=='1' and arr[j+1].second=='1'){
                arr[j].first+=arr[j+1].first;
                arr.erase(arr.begin()+j+1);
                j--;
            }
        }
        for(int j=0;j<num-1;j++){
            if(arr[j].first>arr[j+1].first){
                if(arr[j+1].second=='1' and arr[j].second=='0' ){
                    arr[j+1].second='0';
                    arr[j].second='1';
                }
                // else if(j>0){
                // if(arr[j+1].second=='1' and arr[j].second=='1' and arr[j-1].second=='0' and arr[j-1].first>arr[j+1].first){
                //     arr[j-1].second='1';
                //     arr[j].second = '1';
                //     arr[j+1].second = '0';
                // }
                // }
            }
            
        }
        for(int j=0;j<num;j++){
            if(arr[j].second=='1'){
                result+=arr[j].first;
            }
        }
        cout<<result<<"\n";
    }
    
}

File: session4/137_div2_educational/c.py
from cgi import test


test_num = int(input())
for i in range(test_num) :
    num = int(input())
    arr = [0]*num
    line =input()
    for i in range(len(line)) :
        arr[i] = [0,0]
        arr[i][1] = int(line[i])
    line = list(map(int,input().split(" ")))
    result=0
    for i in range(len(line)) :
        arr[i][0] = line[i]
    for i in range(num-1) :
        if arr[i][1]==1 and arr[i+1][1]==1 :
            arr[i][0] += arr[i+1][0]
            arr.pop(i+1)
    for i in range(num-1) :
        if(arr[i][0]>arr[i+1][0]):
            if arr[i+1][1]==1 and arr[i][1]==0 :
                arr[i][1]=1
                arr[i+1][1]=0
            elif arr[i+1][1]==1 and arr[i][1]==1 and arr[i-1][1]==0 and arr[i-1][0]+arr[i][0]>=arr[i+1][0]+arr[i][0]:
                arr[i-1][1]=1
                arr[i][1]=1
                arr[i+1][1]=0
    for i in range(num) :
        if(arr[i][1]==1) :
            result+=arr[i][0]
    print(result)